os/arch/arm/src/armv7-m/up_mpu.c

```diff
@@ -377,3 +377,40 @@ void up_mpu_set_register(uint32_t *mpu_regs)
 		putreg32(mpu_regs[REG_RASR], MPU_RASR);
 	}
 }
+
+/****************************************************************************
+ * Name: up_mpu_disable_region
+ *
+ * Description:
+ *   Update tcb's mpu register values to disable a mpu region.
+ *   MPU region to be disabled is identified using tcb's mpu registers passed.
+ *   So, if the current MPU h/w settings for this MPU region match to those
+ *   of registers passed, set these new register values to mpu h/w as well.
+ *
+ ****************************************************************************/
+void up_mpu_disable_region(uint32_t *mpu_regs)
+{
+	uint32_t regval;
+
+	/* For kernel task, no operation should be performed. Return immediately */
+	if (!mpu_regs[REG_RBAR]) {
+		return;
+	}
+
+	/* Save original value for mpu_regs[REG_RASR] before proceeding */
+	regval = mpu_regs[REG_RASR];
+
+	/* Disable region's Enable bit in this mpu region settings */
+	mpu_regs[REG_RASR] &= ~MPU_RASR_ENABLE;
+
+	/* Set MPU region number in MPU_RNR register in order to fetch register values */
+	putreg32(mpu_regs[REG_RNR], MPU_RNR);
+
+	/* If original register values match with those of current MPU region
+	 * settings, updated register values should be set so as to disable MPU
+	 * region straight away.
+	 */
+	if (getreg32(MPU_RBAR) == mpu_regs[REG_RBAR] && getreg32(MPU_RASR) == regval) {
```

Comment:
Could you let me know why this conditional is necessary?
Is there any chance to have different values to (with? against?) mpu_regs? Even if it is, why doesn't this work?

Reply:
This condition is needed so that we set updated mpu_regs to mpu h/w only if the mpu_regs passed to this function are the same ones that are set in mpu h/w (meaning that they belong to the same tcb whose mpu_regs are set in mpu h/w).
For example, when there is a fault in a user task, it may happen that before binary manager tries to clean all required tasks and reloads user binary, there may have been context switch to some task other than the one that faulted. In this case, logically, we should disable mpu region for stack protection but we do not need to set updated mpu_regs necessarily.
I think that logically it makes sense that the function that we are providing to disable mpu region sets these updated registers immediately only if required and not always.
For binary recovery upon user task fault, it can perhaps work without these conditions as well.
If as per your understanding these conditions are not necessary, then i will remove them.

Reply:
Removed condition from inside up_mpu_disable_region(..) now as per our discussions.